### PR TITLE
Optimize Regexp_matches to LIKE statements when possible

### PR DESCRIPTION
**Pull request by Tmonster**
As mentioned in the title, this PR is to optimise regexp_matches statements into LIKE statements. LIKE statements will then be optimised into prefix, suffix, or contains statements by the EmptyNeedleRemovalRule.

Of course, I imagine most duckdb users will use `LIKE` by default, but this should help with performance for those who don't. Most optimizations require passing the 's' regexp option, however. Maybe in the future we enable this by default? 

If a user doesn't pass 's', the following transformations won't take place

regex `.` -> LIKE `_`
regex `.*` -> LIKE `%`

There are some corner cases that took some time to think about. Like is directly optimised to contains if only a literal or a string is provided in the regexp argument. If the regexp argument is a concatenation of regex special characters, more attention is paid into whether or not the optimisation should take place. 

If the string or character has any control characters, the optimisation doesn't take place. 
If the regexp has to match any special like characters, then an optimisation to `like_escaped` happens with an escape character of `\`.

Another round of improvements can be made after this merges. We can also optimise case insensitive matches to `ILIKE` and `ILIKE_ESCAPE`. But to avoid a messy large PR, I'll leave that for next week

### CODE REVIEW

File: src/optimizer/rule/regex_optimizations.cpp [hunk 1]
```diff
@@ -12,17 +14,76 @@ namespace duckdb {
...
 	func->function = make_uniq<SpecificFunctionMatcher>("regexp_matches");
-	func->policy = SetMatcher::Policy::ORDERED;
+	func->policy = SetMatcher::Policy::SOME;
 	func->matchers.push_back(make_uniq<ExpressionMatcher>());
 	func->matchers.push_back(make_uniq<ConstantExpressionMatcher>());
+
 	root = std::move(func);
 }
 
+struct LikeString {
```
Comment (Tishj) [inline src/optimizer/rule/regex_optimizations.cpp:24]: very much a nitpick, but I've recently started using unnamed namespaces to essentially make struct definitions static that are only relevant to this file.

```c++
namespace {
struct LikeString {
    bool exists = true;
    string lile_string = "";
};
} //namespace
```

**Review by Tishj — APPROVED** [review]
LGTM

File: src/optimizer/rule/regex_optimizations.cpp [hunk 2]
```diff
@@ -12,17 +14,76 @@ namespace duckdb {
 RegexOptimizationRule::RegexOptimizationRule(ExpressionRewriter &rewriter) : Rule(rewriter) {
 	auto func = make_uniq<FunctionExpressionMatcher>();
 	func->function = make_uniq<SpecificFunctionMatcher>("regexp_matches");
-	func->policy = SetMatcher::Policy::ORDERED;
+	func->policy = SetMatcher::Policy::SOME;
```
Comment (Mytherin) [inline src/optimizer/rule/regex_optimizations.cpp:17]: By changing this to `Some` instead of `Ordered` the order of the matches no longer matters, meaning we could have `[Constant, Expression]` instead of `[Expression, Constant]`. This likely causes some issues with the optimizer below. Maybe we should introduce a `PARTIAL_ORDERED` to accommodate the case here?

File: src/function/scalar/string/like.cpp [hunk 3]
```diff
@@ -522,6 +522,11 @@ void LikeFun::RegisterFunction(BuiltinFunctions &set) {
 	                               nullptr, ILikePropagateStats<NotILikeOperatorASCII>));
 }
 
+ScalarFunction LikeFun::GetFunction() {
+	return ScalarFunction("~~", {LogicalType::VARCHAR, LogicalType::VARCHAR}, LogicalType::BOOLEAN,
```
Comment (Mytherin) [inline src/function/scalar/string/like.cpp:526]: Can we call this function in `LikeFun::RegisterFunction` to avoid duplicated logic?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks good - some comments:

**Comment by Tmonster** [comment]
Forgot that I need to escape LIKE wildcards from literal strings. Should hold off on merging until I get to that tomorrow

File: src/include/duckdb/optimizer/matcher/set_matcher.hpp [hunk 4]
```diff
@@ -79,6 +81,18 @@ class SetMatcher {
 				}
 			}
 			return true;
+		} else if (policy == Policy::PARTIAL_ORDERED) {
+			// partial ordered policy, if too many entries are provided, return false
+			if (matchers.size() < entries.size()) {
```
Comment (Mytherin) [inline src/include/duckdb/optimizer/matcher/set_matcher.hpp:86]: I would imagine this should works the other way around - i.e. we need to match all matchers, but if there are extra expressions they do not count. Otherwise the `bindings` provided to a function are off as they may contain fewer entries than expected.

File: src/optimizer/rule/regex_optimizations.cpp [hunk 5]
```diff
@@ -12,17 +14,137 @@ namespace duckdb {
...
+	} else {
+		auto rune = regexp->rune();
+		if (iscntrl(rune)) {
+			ret.exists = false;
+			return ret;
+		}
+		char chr = toascii(rune);
+		// if a character is equal to the escaped character return that there is no escaped like string.
+		if (!contains && (chr == '%' || chr == '_' || chr == ret.escaped_character[0])) {
+			ret.escaped = true;
```
Comment (Mytherin) [inline src/optimizer/rule/regex_optimizations.cpp:64]: Maybe for simplicity we should just skip the `%` and `_` characters from this optimization for now? It seems like it adds a number of edge cases that complicate this code significantly and I don't think these characters are very common anyway.

File: test/optimizer/regex_optimizer.test [hunk 6]
```diff
@@ -42,3 +41,218 @@ SELECT regexp_matches(s, 'a') FROM test
...
+Binder Error
+
+# when regexp_matches arguments are incorrect, an error is produced
+statement error
+select count(s) from test where regexp_matches('aaa');
+----
+Binder Error
+
+# we escape like special character when we convert to a like string
+query II
```
Comment (Mytherin) [inline test/optimizer/regex_optimizer.test:233]: We are only looking at the plans here but not at the actual results - if we want to leave the escaping code in could we add a bunch of tests that execute the queries and verify the correct results are returned?

**Review by Mytherin — COMMENTED** [review]
Thanks for the fixes! Some more comments:

File: src/include/duckdb/optimizer/matcher/set_matcher.hpp (hunk 4 above)
Comment (Tmonster) [reply src/include/duckdb/optimizer/matcher/set_matcher.hpp:86]: I thought that too, but then how do we match the extra expressions when they are provided? Maybe the FunctionExpressionMatcher can have a minimum_matches variable? And the match function has an extra check like 
```c++
if (entries.size() < minimum_matches) {
    return false;
}
```

Comment (Mytherin) [reply src/include/duckdb/optimizer/matcher/set_matcher.hpp:86]: Ah, I think the way the optimizer works changed since my last review - you do want partial matching on the matchers' side. Perhaps call it `SOME_ORDERED` instead of `PARTIAL_ORDERED` for clarity, as you essentially want `SOME` but ordered?

Comment (Tmonster) [reply src/include/duckdb/optimizer/matcher/set_matcher.hpp:86]: yea. For the regex optimiser to work, I need to match 2 or 3 arguments. In both cases the arguments have to be ordered correctly.
I will change the name to `PARTIAL_ORDERED`. Is the `minimum_matches` variable still a good idea?

Comment (Mytherin) [reply src/include/duckdb/optimizer/matcher/set_matcher.hpp:86]: Fine by me - but then it should apply to some as well

**Comment by Mytherin** [comment]
Thanks!
